# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [13]:
df=pd.read_csv(r'C:\Users\M1043042\Desktop\DSC\Guided Capstone\Data Wrangling\Data-wrangling note books\Step 2\data\step3_output.csv',index_col=0)
df.head()

Name       state  summit_elev  vertical_drop  \
2                Hilltop Ski Area      Alaska         2090            294   
4             Sunrise Park Resort     Arizona        11100           1800   
5   Yosemite Ski & Snowboard Area  California         7800            600   
10               Donner Ski Ranch  California         8012            750   
12                  June Mountain  California        10090           2590   

    trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
2       0        0.0          0          0     0       1  ...   
4       0        0.0          0          1     2       3  ...   
5       0        0.0          0          0     0       1  ...   
10      0        0.0          0          0     0       1  ...   
12      0        0.0          0          2     0       0  ...   

    SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
2                30.0       30.000000        150.000000       36.0   
4               800.0       80.000000        115.000000       49.0   
5                88.0      120.053004        110.000000       84.0   
10              505.0       60.000000        163.000000       82.0   
12             1500.0      120.053004        115.103943       58.0   

    averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
2              69.0     30.000000      34.00000              152.0   
4             250.0     74.000000      78.00000              104.0   
5             300.0     47.000000      47.00000              107.0   
10            400.0     75.000000      75.00000              170.0   
12            250.0     57.916957      64.16681              128.0   

    NightSkiing_ac  clusters  
2             30.0         0  
4             80.0         1  
5              0.0         1  
10             0.0         1  
12             0.0         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [18]:
dummy=pd.get_dummies(df.state)

In [22]:
df=df.drop(columns=['state'])

In [24]:
df = pd.concat([df,dummy], axis=1)

In [53]:
df.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],
      dtype='object')

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [30]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler =preprocessing .StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

D:\anaconda\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
D:\anaconda\lib\site-packages\ipykernel_launcher.py:14: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  


## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [31]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train,y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [35]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [36]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [33]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score,mean_absolute_error

In [47]:
evs=explained_variance_score(y_test,y_pred)
mse=mean_absolute_error(y_test,y_pred)
evs,mse

(-3.0580583332009795e+22, 634474868113.7277)

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [40]:
lm.intercept_

-159568529647.64737

In [42]:
lm.get_params

<bound method BaseEstimator.get_params of LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)>

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [49]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
cof=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
cof.sort_values(by='Coefficient',ascending=False)

Coefficient
fastSixes          2.960343e+13
fastEight          1.230332e+13
trams              8.896465e+12
New York           2.820553e+12
Michigan           2.539768e+12
New Hampshire      2.203359e+12
Pennsylvania       2.107543e+12
Wisconsin          2.107543e+12
Minnesota          1.897012e+12
California         1.897012e+12
Vermont            1.780064e+12
Idaho              1.653161e+12
New Mexico         1.653161e+12
Connecticut        1.513802e+12
Maine              1.513802e+12
Colorado           1.513802e+12
Wyoming            1.513802e+12
Massachusetts      1.358158e+12
Washington         1.358158e+12
Ohio               1.358158e+12
North Carolina     1.179802e+12
West Virginia      1.179802e+12
Illinois           1.179802e+12
Iowa               1.179802e+12
Oregon             1.179802e+12
Utah               1.179802e+12
Montana            1.179802e+12
Virginia           9.662368e+11
Missouri           9.662368e+11
Indiana            9.662368e+11
total_chairs       6.991295e+11
Tennessee          6.852999e+11
Nevada             6.852999e+11
Alaska             6.852999e+11
Arizona            6.852999e+11
South Dakota       6.852999e+11
surface            3.993935e+11
double             3.865063e+11
triple             3.298497e+11
quad               1.994537e+11
fastQuads          1.586111e+11
Maryland           2.819815e+06
New Jersey         4.916162e+04
AdultWeekday       1.014160e+01
Rhode Island       4.441162e+00
summit_elev        1.980233e+00
Runs               1.877686e+00
NightSkiing_ac     1.138672e+00
clusters           9.023438e-01
vertical_drop      8.461690e-01
LongestRun_mi      8.007812e-01
SkiableTerrain_ac  5.537109e-01
yearsOpen          5.488281e-01
Snow Making_ac     5.292969e-01
daysOpenLastYear   4.750977e-01
projectedDaysOpen  4.101562e-01
TerrainParks       3.808594e-01
averageSnowfall    2.861328e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [60]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df.drop(['Name','AdultWeekend', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler =preprocessing .StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

D:\anaconda\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
D:\anaconda\lib\site-packages\ipykernel_launcher.py:21: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.


In [61]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train,y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [62]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred=model.predict(X_test)

In [63]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score,mean_absolute_error

In [64]:
evs=explained_variance_score(y_test,y_pred)
mse=mean_absolute_error(y_test,y_pred)
evs,mse

(0.5476237203815012, 6.927699783353937)

In [65]:
cof=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
cof.sort_values(by='Coefficient',ascending=False)

Coefficient
AdultWeekday       1.089938e+01
summit_elev        3.348269e+00
clusters           2.917163e+00
daysOpenLastYear   2.061170e+00
averageSnowfall    1.877466e+00
vertical_drop      1.529646e+00
Runs               1.426044e+00
triple             1.299687e+00
surface            1.177649e+00
NightSkiing_ac     7.097272e-01
double             6.417389e-01
projectedDaysOpen  6.409033e-01
Snow Making_ac     5.941655e-01
total_chairs       4.089965e-01
quad               4.051694e-01
LongestRun_mi      3.213760e-01
SkiableTerrain_ac  1.949798e-01
yearsOpen          8.947345e-02
TerrainParks       1.312449e-02
fastQuads          7.964157e-03
trams              1.776357e-15
fastEight          1.554312e-15
fastSixes          1.110223e-15

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [72]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df.drop(['Name','AdultWeekend', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming','summit_elev'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler =preprocessing .StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

D:\anaconda\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
D:\anaconda\lib\site-packages\ipykernel_launcher.py:21: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.


In [80]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train,y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [81]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred=model.predict(X_test)

In [82]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score,mean_absolute_error

In [83]:
evs=explained_variance_score(y_test,y_pred)
mse=mean_absolute_error(y_test,y_pred)
evs,mse

(0.5225148504413311, 7.428228597336892)

In [84]:
X.columns

Index(['vertical_drop', 'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad',
       'triple', 'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [85]:
cof=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
cof.sort_values(by='Coefficient',ascending=False)

Coefficient
AdultWeekday       1.058864e+01
averageSnowfall    2.774024e+00
clusters           2.220208e+00
vertical_drop      2.077201e+00
SkiableTerrain_ac  1.847307e+00
daysOpenLastYear   1.475095e+00
triple             1.189496e+00
Runs               9.042282e-01
NightSkiing_ac     8.740295e-01
Snow Making_ac     8.612829e-01
surface            8.248119e-01
double             8.012001e-01
total_chairs       5.821871e-01
quad               2.611841e-01
TerrainParks       2.393553e-01
LongestRun_mi      1.859109e-01
projectedDaysOpen  1.769066e-01
fastQuads          1.114001e-01
yearsOpen          9.393798e-02
trams              1.332268e-15
fastSixes          6.661338e-16
fastEight          4.440892e-16

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -3.0580583332009795e+22 | 634474868113.7277 |-|
| Model 2. | 0.5476237203815012| 6.927699783353937|'state'|
| Model 3. | 0.5225148504413311 | 7.428228597336892|'state','summit_elev','base_elev'|

Model Selection:

Model 2 Suites best since the Explained variance is more and MAE is less and gives an insight that AdultWeekday is feature that management can work on.